### ISSUE TITLE
Use Godot for the user interface and input 

### ISSUE THREAD

**Issue by TechnoPorg** [LMMS/lmms#6194]
### Enhancement Summary
I propose that LMMS should switch to using Godot, rather than QT, for its user interface and input handling. Possibly audio as well, although I'm not sure if Godot's sound system would work well for what LMMS needs. The backend code would become a GDExtension plugin. I would recommend going with 4.0.dev right away, even though it's not stable, to avoid porting everything from 3.x to 4.0 later.
#### Justification
There are several reasons why I believe this to be a good idea:

- QT is massive. v5.15.12 for Windows, which I have installed currently, is 2.28GB. Godot 4.0.dev (https://github.com/godotengine/godot/commit/77721b35ba21f2e7e8bb42cf415fccb018517843) with debug symbols is only 193MB.
- With Godot the UI could be edited with an editor, rather than through code.
- UI/UX elements could be handled in a GDScript, a higher-level language designed specifically for Godot, if desired. QT does have QML, so this point on its own isn't a win for either option.
- However, Godot makes application plugins incredibly simple. All you have to do is load a `.pck` containing a plugin scene, instance said scene, and the plugin is running with full access to the UI elements. This plugin would likely be written in the aforementioned GDScript, which is easy to use.
- It would fix the situation where LMMS has to be installed on Windows in order for the executable to have access to the assets.
- It would be a logical point to implement a whole bunch of big UI changes, notably #1911. If this proposal is accepted I'll gladly try implementing it and overhauling the UI.
#### Mockup
Sorry to inflict my lack of art skill on your eyes :)
![LMMS-Mockup](https://user-images.githubusercontent.com/69441745/136985255-c2529550-e0c2-4467-ae75-8fff8ef73a0b.png)

**Comment by irrenhaus3** [comment]
Pretty interesting idea to use a game engine for the UI layer, although I can imagine a few performance issues arising from the fact that these usually run on a constantly iterating game loop instead of updating on events only. But maybe that's not too much of a concern.
It is true that the QT Widgets library is rather old and makes UI design more cumbersome than modern app development is used to, and LMMS would definitely benefit from adaptable frontends in the long term.
Take note though that replacing the frontend implies a massive refactor that for now is very likely unfeasible. There is an idea to separate frontend and "business logic" on the binary level by moving core functionality into a library and implement the UI in the actual executable file, that then calls into the core library. We're still pretty far from being able to do that, however.  
  
Long story short: Interesting idea, but probably not possible right now.

**Comment by TechnoPorg** [comment]
> Do you think it would be possible to make some port to Godot, I'm a godot user, so maybe I can help with that. Even though I can't code with C++

@ca3games A port to Godot will absolutely be possible in the future, but not with how the UI stuff is set up right now. @irrenhaus3 you had said that core/frontend elements need to be decoupled before any major UI changes can be made, is there any way I can help with that?

> What about using libSweell ? Guys, Reaper it's just 100Mb (a bit more).

@Daniele71 I feel that Swell isn't the best option to make a frontend for LMMS, because as far as I can tell there's no editor, just C++ code, which is pretty similar to QT. Also not a lot of people have ever heard of it, me included before today, which makes it more intimidating for new contributors. Lots of people know Godot, and thus would more easily be able to contribute.

**Comment by SeleDreams** [comment]
I think one advantage of using Godot could be thanks to the huge community of Godot Engine, I think improvements to the Godot Engine could benefit LMMS, while improvements to LMMS could benefit the Godot Engine, as often Godot developers want to improve the workflow of open source asset creation tools with Godot

**Comment by Gabrielxd195** [comment]
I don't know anything about programming but I love godot, but researching about the subject I found this: 

https://kivy.org/#home 

I don't know guys I got excited, but maybe this could be of help.

**Comment by SeleDreams** [comment]
> I don't know anything about programming but I love godot, but researching about the subject I found this:
> 
> https://kivy.org/#home
> 
> I don't know guys I got excited, but maybe this could be of help.

It's python, this isn't appropriate.

Godot although it has support for higher level languages is fully programmed in C++

**Comment by rdrpenguin04** [comment]
I mean, a potential UI could be programmed in PERL for all we care. Python isn't necessarily a bad choice, but there should be *at least* one UI available in C++.

Probably. I'm not an expert here.

**Comment by SeleDreams** [comment]
> I mean, a potential UI could be programmed in PERL for all we care. Python isn't necessarily a bad choice, but there should be _at least_ one UI available in C++.
> 
> Probably. I'm not an expert here.

I could understand an UI in C# or co, but python is a very high level language with dynamic typing, interfacing with the C++ source would be a pain and would lead to performance issues

**Comment by RiedleroD** [comment]
If LMMS switched to electron, I'd have to switch to a different DAW. Discord is already hardly navigatable for me, and I think using electron in general would be a massive downgrade to the current Qt5 GUI.

**Comment by wwderw** [comment]
As someone that does use Godot for regular application developed, it does make for a stable SDK for such.

Keep in mind, Godot UI itself, is also using the engine (eating one's own food so to speak).

There is a low processor mode, it does allow for manual memory management as well.  Just try to stay away from delta time type of coding (or make sure to manually handle the memory) and it should be good.

Can use other langues (C++ and even Rust, I like Nim myself, but I digress) as well if don't want to use GDScript as much.

Godot will still be less resource heavy compared to Electron, so keep that in mind as well.  Pixelorama and Material Maker are good projects that are made using the Godot Engine (and Pixelorama to my knowledge is still all GDScript).

**Comment by wwderw** [comment]
> What I mean by that is that Electron's resource consumption depends on what you are actually doing. Electron takes fewer resources if you compare the power required to run VisualStudio (based on the Electron) and Godot, you will find that Visual takes several orders of magnitude in fewer resources than Godot.

I don't know about this. And it also may depend on the host platform  I think it all depends on what you are using within VSCode as well versus Godot.  On Linux, I see far significant less resource consumption with Godot versus any Electron app that I have created or have used created by people that probably did things better compared to myself (coding wise).

I could see more with using Webview compared to full blown Electron.  Node and even customized Chromium is a lot of resource usage before getting to the app itself.  While I am not a fan of GTK and personally prefer Qt over it, Webview would, in my opinion, be better compared to full on Electron.  Now the one argument that you could have with Electron versus Webview is that don't have to do much CSS/SVG work to make sure everything renders the same on all platforms.  It's not to bad to do that with Webview with SVGs and use those instead of the default renders of the platform.

Using Webkit (not Electron) over Godot, maybe, it all depends on what one is trying to do.

Using web-browsers (be it stripped down webview or custom chromium/node) are trying to shoehorn a browser environment for a desktop application.

Now immediate mode gui isn't so much the issue that a lot of people think that it is.  Even using Raygui (a module of Raylib) with C++, could see roughly the same as Qt (depending on the complexity of the app).  Shoot Blender is immediate mode with openGL.

**Comment by SeleDreams** [comment]
this comparison doesn't make sense, of course is we make a big 3d game it
would take more resources

a big 3d game in godot tho would still be more efficient than a big 3d game
in JS

**Comment by rdrpenguin04** [comment]
First of all, GDScript is remarkably fast.
Second of all, you don't have to use GDScript with Godot; there are other language options.
JS does not mix well with high performance audio applications. I would know; I've tried to make an audio worklet, and it's incredibly slow.
I would not recommend Electron even remotely for LMMS. Nevertheless, once we abstract the UI out, it would be possible.

**Comment by SeleDreams** [comment]
> 1: JS code is faster than GDScript. (via Google v8 engine). 2: Just because a game is encoded in an interpreted language doesn't mean that the game in question will be absolutely slow because JS or GDScript code only calls APIs that are usually written in C ++. 3: If you are not convinced then tell me why use Godot with GDScript when it does not match the speed of C ++? Why Blender uses Python to create 3D APIs and ADDON? Not all of these languages ​​are very fast, yet they do the job they are asked to do. And I remind you that we can use WASM for the critical parts of a Web application.
> 
> Electron is a great fit for the LMMS and there is already music design software like LMMS that works great on the Electron. (and more generally directly in the web browser)

technically you can program in godot using nothing but C++, that's often what I do
the main issue of web based frameworks is that they're heavy and i don't talk about the code of the app itself, the web browser in itself is heavy, they're often based on chrome, and chrome has a lot of ram usage 
godot at least can directly communicate with the GPU and CPU rather than going through a sandbox (web based stuff is always sandboxed for security)
another issue is JS, JS is just not a good language at all and lots of js programmers program like shit, making lmms JS would only bring lots of bad code

**Comment by SeleDreams** [comment]
What's sure at least is I wouldn't contribute code to a JS/browser based project as it's just too much of a mess

**Comment by SeleDreams** [comment]
godot wasn't suggested because it was "less complicated" that was what some users thought and it's subjective but more because of its advantages of being lightweight, having a wide support of languages, having a nice built in UI system, and being modular

also it doesn't require many dependencies from what I remember

**Comment by SeleDreams** [comment]
the other advantage is godot is another open source project so both LMMS and Godot contributors could help both projects at once

**Comment by SeleDreams** [comment]
electron is open source, but it's more company based than community based, godot is entirely supported by volunteers like LMMS

**Comment by wwderw** [comment]
> Electron exposes features that are not directly available in web browsers, which allows direct communication with the user's operating system. So no sandbox problem

Thru Node, yes.  Webkit also allows for atleast some of this as well.  At least the last time that I tried messing with this type of SDK.  There are usually some module that one can import to do your own server to use via webkit to take the place what node does as well.

As long as it is using some type of browser (be it webkit or Electron) it is still going to not quite be the same as traditional apps.  The browser experience is just following a different paradigm.  I think that they could do better to help bridge that experience, but that remains to be seen if the powers that be go down that route.

Now using something like Raygui (although that may not be feature complete for a complex app like this) with C/C++ would yield better resource usage compared to Godot.  Unless minimize how much is done via GDScript and do the logic totally in language like C++.

**Comment by SeleDreams** [comment]
> > Electron exposes features that are not directly available in web browsers, which allows direct communication with the user's operating system. So no sandbox problem
> 
> Thru Node, yes. Webkit also allows for atleast some of this as well. At least the last time that I tried messing with this type of SDK. There are usually some module that one can import to do your own server to use via webkit to take the place what node does as well.
> 
> As long as it is using some type of browser (be it webkit or Electron) it is still going to not quite be the same as traditional apps. The browser experience is just following a different paradigm. I think that they could do better to help bridge that experience, but that remains to be seen if the powers that be go down that route.
> 
> Now using something like Raygui (although that may not be feature complete for a complex app like this) with C/C++ would yield better resource usage compared to Godot. Unless minimize how much is done via GDScript and do the logic totally in language like C++.

I have just a minimal experience with raylib stuff, always had some interest but never the time to look at it. I dunno if its UI system is as good as godot's one tho

**Comment by rdrpenguin04** [comment]
I do in fact have legitimate uses for Electron; primarily, I'm making an IDE for a school-related project, and Electron has a good history with small IDEs.

Electron is powerful, but it is slow, and I have seen that firsthand. Real-time audio synthesis needs maximum speed with minimum compromises, and that is something a browser engine can not provide, not with the JS engine or even the DOM engine.

**Comment by rdrpenguin04** [comment]
LMMS has been dealing with Qt breaks for quite a while; also, Godot breaks are so minor that they can get away with putting documentation for 4.x into the 3.x docs...
No, they can't get away with it, but the language is apparently similar enough that it'll stand up. My point is that breaking changes won't be an issue.
Secondly, you seem to have forgotten that the DOM engine is just as much of a problem as the JS engine is, and it's not something you can simply circumvent.

**Comment by wwderw** [comment]
> I have just a minimal experience with raylib stuff, always had some interest but never the time to look at it. I dunno if its UI system is as good as godot's one tho

No, Godot's predefined widgets/nodes is better, in my opinion, compared to Raygui.  Simple apps, Raygui does well.

There is, to tangentially go with this project uses, a program reminiscent of JFXR (JFXR being browser based), that was made using Raylib, by the creator of Raylib.

I think immediate mode gui gets a bad rap, and I personally like them and done right, not that much of a performance hit compared to the traditional cross platform toolkits.  

I've dabbled using AudioNodes(Electron based), and while I didn't have a problem using it and I did like its feature set, it did like its resources.

**Comment by rdrpenguin04** [comment]
> I don't see where the problem is with the DOM engine. In Godot, the UI is also managed as a tree like the DOM.

The tree isn't the problem, the problem is the layout and styling required for a DOM to be rendered. Godot is optimized to handle simple information simply. Electron is optimized to handle complex information in whatever way necessary, which makes it significantly slower.

> When interacting with Visual Studio and Discord, I don't have a problem with slowness anyway.

Just because you have no issues, that doesn't mean someone else wouldn't have the same problems. I have often been on the side of slower than average, so I have plenty of experience in this area.

**Comment by Veratil** [comment]
I believe you're all missing the point of what we're trying to get LMMS to. If we get the core split from Qt then we have a DAW backend as `liblmms` essentially. The frontend after that can be in Qt, Electron, or Godot, or anything else really.

That being said, I believe we'll keep the frontend as Qt for now since most of the work for it is there already and updating is fairly quick to new Qt versions. What we get out of the core separation though is the ability for _all of you_ to create these frontends to expand the reach of LMMS. Electron could allow people to run a version of LMMS in the browser on the go. Qt would be the desktop version. Godot could be an alternative and possibly after some time and learning the default desktop option. It would definitely be able to draw faster using the GPU than Qt afaik. :)

**Comment by wwderw** [comment]
> When interacting with Visual Studio and Discord, I don't have a problem with slowness anyway.

Depending on the specs of your machine, may never see an issue.  Doesn't mean the individual program isn't resource hungry though.

**Comment by rdrpenguin04** [comment]
> I believe you're all missing the point of what we're trying to get LMMS to. If we get the core split from Qt then we have a DAW backend as `liblmms` essentially. The frontend after that can be in Qt, Electron, or Godot, or anything else really.

For the record, I did address that:

> Nevertheless, once we abstract the UI out, it would be possible.

**Comment by radlotus** [comment]
Guys, I just wanna say that I really like the very easy piano roll and pattern system, and the browser on the left. Out of Ableton, Reaper, Ardour and FL Studio I like LMMS the most. It's very easy to write melodies and work with patterns and introduce complexities and variations. FL Studio is the closest, but i really hate the convoluted pattern system and the floating windows management which i think its worse than LMMS. I don't mind the floating windows in LMMS. I think LMMS's interface shouldnt deviate too much. I think its fine right now just as it is that's why its been successful even with so many drawbacks, that mostly consist of absence of some features. I don't want LMMS to become in terms of UI like Reaper, Ableton, Zrythm, Waveform, Cakewalk, Studio One, Pro Tools which share very similar ui which i don't like, except maybe for Reaper. I like Reaper the most out of the listed. It's easy actually and very stable and functional, but ultimately I like LMMS more because its easier on my adhd brain and the music i make is heavily based on complex beats and melodies which i write by hand in the piano roll.

**Comment by aaronfranke** [comment]
Disclaimer: I actively contribute to Godot Engine.

Godot is indeed a really good GUI framework. GDScript, Control nodes, and Godot's visual editor makes it super easy. Godot also has a lighter footprint than most toolkits, even Electron. Note that Godot has one major disadvantage vs others which is that it requires OpenGL or Vulkan to render things, so it won't work well in VMs or really old systems.

> I would recommend going with 4.0.dev right away, even though it's not stable, to avoid porting everything from 3.x to 4.0 later.

4.0 dev is very unstable right now, so I'd be hesitant to suggest this, but maybe it's a good choice in the long term. Considering this issue is still being debated and there needs to be core refactoring done to split the front-end from the back-end, it may be better to wait.

> although I can imagine a few performance issues arising from the fact that these usually run on a constantly iterating game loop instead of updating on events only. But maybe that's not too much of a concern.

Godot has mechanisms for event-driven programming including signals and notifications, it's completely optional to use every-frame logic like the `_process` function. Also note that Godot has a low processor mode for not updating when nothing changes on the screen, which the editor uses. However when something does need to update, Godot does not have partial redraws, so it has to redraw everything.

> And Qt is used for so much more than just GUI. It provides memory cleanup and the signal/slot mechanism. Plus some other silly stuff like XML, I/O, timers, smart pointers, threads, vectors, string translation and debug messages. All of this can be done in some other way but someone's got to do it.

I think Godot provides all of this (not sure about smart pointers).

> I haven't found any supports in Godot even for multi-window. 

Godot 4.0 will have support for multiple windows.

> Additionally Godot already has a mechanism in which you can build the engine with a selection of components in order to reduce the executable size. Currently there is no mode for the UI only AFAIK but that can possibly be added.

I plan to make an unofficial GUI-optimized fork of Godot that will be released alongside Godot 4.0, but I haven't made the repository/branch yet because I am trying to get most of the heavy lifting work upstreamed into Godot.

**Comment by Yam76** [comment]
So if I understand correctly, this project's future work is to extract lmms from Qt into `liblmms`, then try out Godot 4.0 when it arrives, and if it works well, great, if it doesn't, stick with Qt.

**Comment by Veratil** [comment]
> So if I understand correctly, this project's future work is to extract lmms from Qt into `liblmms`, then try out Godot 4.0 when it arrives, and if it works well, great, if it doesn't, stick with Qt.

The plan is to split the core and the frontend, period. This allows us to create frontends for anything: mobile, TUI, a MIDI tracker (with expanded capabilities of a full DAW), etc. We could have a frontend with Qt and have a frontend with Godot. We don't have to remove one to have the other with the split setup.

**Comment by Yam76** [comment]
Will there still be an official frontend? I'm mentioning this because I am concerned it will turn into a massive mess otherwise. Poor new user experience, messy and abandoned frontends, a growing disconnect between what backend devs write and what frontend devs want, etc.

**Comment by Veratil** [comment]
> Will there still be an official frontend? I'm mentioning this because I am concerned it will turn into a massive mess otherwise. Poor new user experience, messy and abandoned frontends, a growing disconnect between what backend devs write and what frontend devs want, etc.

Yes, we'll always have an officially supported frontend for desktop. Which framework ends up being is up to community consensus and *ease of work*. Qt will be the main one for now, and if Godot proves to be easier to maintain and extend as needed then it's entirely possible to move to it. A good argument will be needed to completely switch to a new framework though since Qt is dug in deep right now. And remember it's not just me you have to convince. :) (personally I'm a fan of Godot and have been wanting to learn it, so this would be a good excuse to finally jump in)

As for messy and abandoned frontends, I will always fight against having multiple frontends in the main repo unless they are going to be actively supported. That being said, we're still a good ways away from getting to this point, but a well thought out plan is a good thing.

**Comment by wwderw** [comment]
Could always think of going the CopperSpice route as well.  For those that don't know, it's a fork of Qt when it was still LGPL/GPL2.

I'm a big fan of Godot and I do like immediate mode GUIs, I use Fyne with Golang as well (I'm not saying go with Fyne/Golang at all, just mentioning it because it's an immediate mode GUI).

**Comment by Spekular** [comment]
Locking this for now, because I don't think it's particularly productive to continue this discussion right now. Until the core and UI are properly separated a Godot UI isn't happening, and once they are:

- The main frontend should probably be decided via a discussion that doesn't start at the conclusion. Think "what framework should we use?", not "why should we use framework x?"
- Anyone who wants to can create a separate frontend with Godot / Electron / Vulkan / Etch-a-sketch

I also feel that these kinds of issues generate lots of:

- Unfounded excitement: "it'll be awesome when [my specific vision of the future] happens"
- Unfounded worry: "please don't remove the piano roll when switching frameworks"
- Conversation rehashing things that have already been decided or thoroughly discussed: "what if the core was separate from the frontend?"

**Comment by tresf** [comment]
> Locking this for now, because I don't think it's particularly productive to continue this discussion right now. Until the core and UI are properly separated a Godot UI isn't happening, and once they are:

Might be a better candidate for Discussions.  (Do people use those?) https://github.com/LMMS/lmms/discussions.  I feel that there's some good information here for people looking to prototype something, especially when we have active framework developers chiming in. 🍻

**Comment by tresf** [comment]
> Doubting whether Godot is a good choice for such a big application... I was eagerly looking for a flexible UI layout with the docking panels, like Visual Studio and Eclipse and also mentioned in an issue 6 years ago here: [#1911 (comment)](https://github.com/LMMS/lmms/issues/1911#issuecomment-88131481) I consider the docking should be a basic UI feature for big applications - even VC++6.0 did this better than most of others - and with Qt this mightn't be too difficult since we have QDockWidget

Interestingly, if you read the [fourth comment](https://github.com/LMMS/lmms/issues/1911#issuecomment-88137166) on that same bug report, you have the author of ZynAddSubFX making the following statement about Qt:

> Well, QT pretty much anything is very much unsafe to use anywhere near the realtime side of things.

... a comment about whether or not the signals and slots system offered by Qt is a suitable long-term solution for core/ui communication.
